service/history/workflow/update/update.go

```diff
@@ -156,34 +160,96 @@ func newCompleted(
 	return upd
 }
 
+// WaitLifecycleStage waits until the Update has reached at least `waitStage` or
+// a timeout. If the Update reaches `waitStage` with no timeout, the most
+// advanced stage known to have been reached is returned, along with the outcome
+// if any. If there is a timeout due to the supplied soft timeout, then
+// unspecified stage and nil outcome are returned, without an error. If there is
+// a timeout due to context deadline expiry, then the error is returned as usual.
+func (u *Update) WaitLifecycleStage(
+	ctx context.Context,
+	waitStage enumspb.UpdateWorkflowExecutionLifecycleStage,
+	softTimeout time.Duration) (stage enumspb.UpdateWorkflowExecutionLifecycleStage, outcome *updatepb.Outcome, err error) {
+
+	stage = enumspb.UPDATE_WORKFLOW_EXECUTION_LIFECYCLE_STAGE_UNSPECIFIED
+	switch waitStage {
+	case enumspb.UPDATE_WORKFLOW_EXECUTION_LIFECYCLE_STAGE_ACCEPTED:
+		return u.waitLifecycleStage(ctx, u.WaitAccepted, softTimeout)
+	case enumspb.UPDATE_WORKFLOW_EXECUTION_LIFECYCLE_STAGE_COMPLETED:
+		return u.waitLifecycleStage(ctx, u.WaitOutcome, softTimeout)
+	default:
+		return enumspb.UPDATE_WORKFLOW_EXECUTION_LIFECYCLE_STAGE_UNSPECIFIED, nil, serviceerror.NewUnimplemented(
+			fmt.Sprintf("%v is not implemented", waitStage))
+	}
+}
+
+func (u *Update) waitLifecycleStage(
+	ctx context.Context,
+	waitFn func(ctx context.Context) (enumspb.UpdateWorkflowExecutionLifecycleStage, *updatepb.Outcome, error),
+	softTimeout time.Duration) (enumspb.UpdateWorkflowExecutionLifecycleStage, *updatepb.Outcome, error) {
+
+	type result struct {
+		stage   enumspb.UpdateWorkflowExecutionLifecycleStage
+		outcome *updatepb.Outcome
+		err     error
+	}
+	ch := make(chan result, 1)
+
+	go func() {
+		ctx, cancel := context.WithTimeout(context.Background(), softTimeout)
+		defer cancel()
+		stage, outcome, err := waitFn(ctx)
+		if err != nil && common.IsContextDeadlineExceededErr(err) {
```

Comment:
You shouldn't care about the returned error, the context deadline error might be swallowed internally.

```suggestion
		if ctx.Error() != nil {
```